rskj-core/src/main/java/co/rsk/db/importer/provider/BootstrapFileHandler.java

```diff
@@ -0,0 +1,129 @@
+            BootstrapURLProvider bootstrapUrlProvider,
+            Unzipper unzipper) {
+        this.bootstrapUrlProvider = bootstrapUrlProvider;
+        this.unzipper = unzipper;
+    }
+
+    // for testing purposes these 3 methods are splitted into 3 public items
+    public void setTempDirectory() {
+        this.tempPath = getTempDirectory();
+    }
+
+    public byte[] getBootstrapData() {
+        Path bootstrapPath = tempPath.resolve(EXTRACTED_PATH);
+        try {
+            return Files.readAllBytes(bootstrapPath);
+        } catch (IOException e) {
+            throw new BootstrapImportException(String.format(
+                    "Failed reading data after decompression %s", bootstrapPath), e);
+
+        }
+    }
+
+    public void retrieveAndUnpack(Map<String, BootstrapDataEntry> selectedEntries) {
+        Path zipPath = tempPath.resolve(BOOTSTRAP_DATA_NAME);
+
+        List<String> publicKeys = new ArrayList<>(selectedEntries.keySet());
+        String publicKey = publicKeys.get(new SecureRandom().nextInt(publicKeys.size()));
+        String expectedHash = selectedEntries.get(publicKey).getHash();
+
+        String dbSuffix = selectedEntries.get(publicKey).getDb();
+        URL dbURL = bootstrapUrlProvider.getFullURL(dbSuffix);
+
+        downloadFile(zipPath, dbURL);
+        checkFileHash(zipPath, expectedHash);
+        unzipFile(zipPath, tempPath);
+    }
+
+    private void checkFileHash(Path databasePath, String expectedHash) {
+        try {
+            byte[] fileContent = Files.readAllBytes(databasePath);
+            byte[] hash = HashUtil.sha256(fileContent);
+            if (!Arrays.equals(hash, Hex.decode(expectedHash))) {
+                throw new BootstrapImportException(String.format(
+                        "File: %s does not match with expected hash: %s", databasePath, expectedHash));
+            }
+            logger.info("Bootstrap data hash checked");
+        } catch (IOException e) {
+            throw new BootstrapImportException(
+                    "Error trying to read bootstrap data contents. Please start again the import process", e);
+        }
+    }
+
+    private void unzipFile(Path sourcePath, Path destinationPath) {
+        try (InputStream inputStream = Files.newInputStream(sourcePath)){
+            unzipper.unzip(inputStream, destinationPath);
+            logger.info("Bootstrap data extracted");
+        } catch (IOException e) {
+            throw new BootstrapImportException(
+                    "The file is corrupted or incomplete. Please start again the import process", e);
+        }
+    }
+
+    private Path getTempDirectory() {
+        try {
+            return Files.createTempDirectory("import");
+        } catch (IOException e) {
+            throw new BootstrapImportException(
+                    "Failed to create a temporary directory. Please start again the import process", e);
+        }
+    }
+
+    private void downloadFile(Path downloadPath, URL url) {
```

Comment:
we can implement a timeout of some kind here, taking advantage of the steam option. assuming someone has push access to s3, they could create arbitrary random signatures for non-existent data, and we would be able to check it just when the file has been fully downloaded. this could lead to a DoS. 

when we change to a more permission-less structure and ditch the authentication-to-upload part, we should look into better mitigations for this.